bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/JsonParser.java

```diff
@@ -101,1172 +130,1049 @@ public static Object parse(InputStream in, String charsetName) throws BError {
+                return null;
             }
-
+            throw createError(ErrorReasons.BALLERINA_PREFIXED_CONVERSION_ERROR,
+                    ErrorHelper.getErrorDetails(ErrorCodes.CANNOT_CONVERT_NIL, targetType));
         }
 
-        /**
-         * Represents the state where an escaped character is processed in a field name.
-         */
-        private static class FieldNameEscapedCharacterProcessingState extends EscapedCharacterProcessingState {
-
-            @Override
-            protected State getSourceState() {
-                return FIELD_NAME_STATE;
+        private static Object convertArray(ArrayValueImpl array, Type targetType, Type targetRefType,
+                                           Set<TypeValuePair> unresolvedValues) {
+            switch (targetType.getTag()) {
+                case TypeTags.ARRAY_TAG:
+                    ArrayType arrayType = (ArrayType) targetType;
+                    ArrayValueImpl newArray = new ArrayValueImpl(targetRefType, arrayType.getSize());
+                    for (int i = 0; i < array.size(); i++) {
+                        newArray.addRefValueForcefully(i, convert(array.getRefValue(i), arrayType.getElementType(),
+                                unresolvedValues));
+                    }
+                    return newArray;
+                case TypeTags.TUPLE_TAG:
+                    TupleType tupleType = (TupleType) targetType;
+                    int minLen = tupleType.getTupleTypes().size();
+                    BListInitialValueEntry[] tupleValues = new BListInitialValueEntry[array.size()];
+                    for (int i = 0; i < array.size(); i++) {
+                        Type elementType = (i < minLen) ? tupleType.getTupleTypes().get(i) : tupleType.getRestType();
+                        tupleValues[i] = ValueCreator.createListInitialValueEntry(convert(array.getRefValue(i),
+                                elementType, unresolvedValues));
+                    }
+                    return new TupleValueImpl(targetRefType, tupleValues);
+                case TypeTags.TABLE_TAG:
+                    TableType tableType = (TableType) targetType;
+                    for (int i = 0; i < array.size(); i++) {
+                        Object bMap = convert(array.get(i), tableType.getConstrainedType(), unresolvedValues);
+                        array.setRefValueForcefully(i, bMap);
+                    }
+                    array.setArrayRefTypeForcefully(TypeCreator.createArrayType(tableType.getConstrainedType()),
+                            array.size());
+                    BArray fieldNames = StringUtils.fromStringArray(tableType.getFieldNames());
+                    return new TableValueImpl<>(targetRefType, array, (ArrayValue) fieldNames);
+                default:
+                    break;
             }
-
+            // should never reach here
+            throw createConversionError(array, targetType);
```

Comment:
Can we move this throw statement to the default case? If possible check other occurrences as well.